components/search/src/types.rs

```diff
@@ -0,0 +1,197 @@
+
+    /// The distribution id for the user's build.
+    pub distribution_id: String,
+
+    /// The search related experiment id that the user is included within. On
+    /// desktop this is the `searchConfiguration.experiment` variable.
+    pub experiment: String,
+
+    /// The application name that the user is using.
+    pub app_name: SearchApplicationName,
+
+    /// The application version that the user is using.
+    pub version: String,
+}
+
+/// Parameter definitions for search engine URLs. The name property is always
+/// specified, along with one of value, experiment_config or search_access_point.
+#[derive(Debug, uniffi::Record)]
+pub struct SearchUrlParam {
+    /// The name of the parameter in the url.
+    pub name: String,
+    /// The parameter value, this may be a static value, or additionally contain
+    /// a parameter replacement, e.g. `{inputEncoding}`. For the partner code
+    /// parameter, this field should be `{partnerCode}`.
+    pub value: Option<String>,
+    /// The value for the parameter will be derived from the equivalent experiment
+    /// configuration value.
+    /// Only desktop uses this currently.
+    pub experiment_config: Option<String>,
+}
+
+/// Defines an individual search engine URL.
+#[derive(Debug, uniffi::Record)]
+pub struct SearchEngineUrl {
+    /// The PrePath and FilePath of the URL. May include variables for engines
+    /// which have a variable FilePath, e.g. `{searchTerm}` for when a search
+    /// term is within the path of the url.
+    pub base: String,
+
+    /// The HTTP method to use to send the request (`GET` or `POST`).
+    /// If not specified, defaults to GET.
+    pub method: Option<String>,
+
+    /// The parameters for this URL.
+    pub params: Option<Vec<SearchUrlParam>>,
+
+    /// The name of the query parameter for the search term. Automatically
+    /// appended to the end of the query. This may be skipped if `{searchTerm}`
+    /// is included in the base.
+    pub search_term_param_name: Option<String>,
+}
+
+/// The URLs associated with the search engine.
+#[derive(Debug, uniffi::Record)]
+pub struct SearchEngineUrls {
+    /// The URL to use for searches.
+    pub search: SearchEngineUrl,
+
+    /// The URL to use for suggestions.
+    pub suggestions: Option<SearchEngineUrl>,
+
+    /// The URL to use for trending suggestions.
+    pub trending: Option<SearchEngineUrl>,
+}
+
+/// A definition for an individual search engine to be presented to the user.
+#[derive(Debug, uniffi::Record)]
+pub struct SearchEngineDefinition {
+    /// An optional list of aliases for this engine.
+    pub aliases: Option<Vec<String>>,
+
+    /// The classification of search engine according to the main search types
+    /// (e.g. general, shopping, travel, dictionary). Currently, only marking as
+    /// a general search engine is supported.
+    /// On Android, only general search engines may be selected as "default"
+    /// search engines.
+    pub classification: String,
```

Comment:
Do you think this will become an enum eventually, or is the idea to keep it as flexible as possible?

Reply:
I'll make it an enum, to reflect what we currently have in the schema. It might become interesting for what we do when processing the search configuration, but I'll be looking at that in the next patch.